src/Storages/S3Queue/S3QueueSource.cpp

```diff
@@ -42,33 +43,112 @@ StorageS3QueueSource::S3QueueKeyWithInfo::S3QueueKeyWithInfo(
 StorageS3QueueSource::FileIterator::FileIterator(
     std::shared_ptr<S3QueueFilesMetadata> metadata_,
     std::unique_ptr<GlobIterator> glob_iterator_,
+    size_t current_shard_,
     std::atomic<bool> & shutdown_called_)
     : metadata(metadata_)
     , glob_iterator(std::move(glob_iterator_))
     , shutdown_called(shutdown_called_)
+    , log(&Poco::Logger::get("StorageS3QueueSource"))
+    , sharded_processing(metadata->isShardedProcessing())
+    , current_shard(current_shard_)
 {
+    if (sharded_processing)
+    {
+        for (const auto & id : metadata->getProcessingIdsForShard(current_shard))
```

Comment:
maybe try out [std::ranges::iota_view](https://en.cppreference.com/w/cpp/ranges/iota_view) to avoid allocating `std::vector`?
I know it's not a performance critical place but it's more a fun fact if you are interested 😄